File: lc_soln/1944_number_of_visible_people_in_queue/solution.py
from typing import List


class Solution:
    def canSeePersonsCount(self, heights: List[int]) -> List[int]:
        n = len(heights)
        result = [0] * n
        stack = []
        for i in range(n - 1, -1, -1):
            while stack and stack[-1] <= heights[i]:
                result[i] += 1
                stack.pop()
            if stack:
                result[i] += 1
            stack.append(heights[i])
        return result

File: lc_soln/1944_number_of_visible_people_in_queue/solution.cpp
class Solution {
public:
    vector<int> canSeePersonsCount(vector<int>& heights) {
        int n = heights.size();
        vector<int> result(n, 0);

        vector<int> stack;
        for (int i = n - 1; i >= 0; --i) {
            while (!stack.empty() && heights[stack.back()] <= heights[i]) {
                result[i]++;
                stack.pop_back();
            }
            if (!stack.empty()) {
                result[i]++;
            }
            stack.push_back(i);
        }
        return result;
    }
};

